src/api/messages/getSingleMessage.js

```diff
@@ -0,0 +1,133 @@
+import { AvatarURL } from '../../utils/avatar';
+
+/**
+ * The variant of `Reaction` found in the actual server response.
+ *
+ * Note that reaction events have a *different* variation; see their
+ * handling in `eventToAction`.
+ */
+export type ServerReaction = {|
+  +emoji_name: string,
+  +emoji_code: string,
+  +reaction_type: ReactionType,
+
+  // Our binding assumes FL 120 (see jsdoc), so we don't have to handle FLs
+  // <2 that don't provide `user_id`.
+  +user_id: UserId,
+
+  // Deprecated and "to be removed in a future release":
+  //   https://zulip.com/api/get-message
+  -user: $ReadOnly<{|
+    id: UserId,
+    email: string,
+    full_name: string,
+
+    // Not observed, so marked optional (doc says required)
+    is_mirror_dummy?: boolean,
+  |}>,
+|};
+
+/**
+ * The elements of Message.edit_history found in the actual server response.
+ */
+export type ServerMessageEdit = {|
+  // Our binding assumes FL 120 (see jsdoc), so we don't have to support
+  // the pre-FL 118 shape. This type is written accordingly.
+
+  +prev_content?: string,
+  +prev_rendered_content?: string,
+  +prev_rendered_content_version?: number,
+  +prev_stream?: number,
+  +prev_topic?: string,
+  +stream?: number,
+  +timestamp: number,
+  +topic?: string,
+  +user_id: UserId | null,
+|};
+
+// How `ServerMessage` relates to `Message`, in a way that applies
+// uniformly to `Message`'s subtypes.
+type ServerMessageOf<M: Message> = {|
+  ...$Exact<M>,
+
+  // Can be null. The server hard-codes client_gravatar:
+  //   https://chat.zulip.org/#narrow/stream/378-api-design/topic/.60client_gravatar.60.20in.20.60messages.2F.7Bmessage_id.7D.60/near/1418337
+  +avatar_url: string | null,
+
+  +reactions: $ReadOnlyArray<ServerReaction>,
+
+  // Unlike Message['edit_history'], this can't be `null`.
+  +edit_history?: $ReadOnlyArray<ServerMessageEdit>,
+|};
+
+export type ServerMessage = ServerMessageOf<PmMessage> | ServerMessageOf<StreamMessage>;
+
+// The actual response from the server.  We convert the message to a proper
+// Message before returning it to application code.
+type ServerApiResponseSingleMessage = {|
+  ...$Exact<ApiResponseSuccess>,
+  -raw_content: string, // deprecated
+  +message: ServerMessage,
+|};
+
+/** Exported for tests only. */
+export const migrateMessage = (
+  message: ServerMessage,
+  identity: Identity,
+  allowEditHistory: boolean,
+): Message => ({
+  ...message,
+  avatar_url: AvatarURL.fromUserOrBotData({
+    rawAvatarUrl: message.avatar_url,
+    email: message.sender_email,
+    userId: message.sender_id,
+    realm: identity.realm,
+  }),
+  reactions: message.reactions.map(reaction => {
+    const { user_id, emoji_name, reaction_type, emoji_code } = reaction;
+    return { user_id, emoji_name, reaction_type, emoji_code };
+  }),
+
+  // Why condition on allowEditHistory? See MessageBase['edit_history'].
+  edit_history: allowEditHistory
+    ? (message.edit_history: $ReadOnlyArray<MessageEdit> | void)
+    : null,
+});
```

Comment:
For how we type the data we expect from the server—`ServerMessage`—I think we first want to make it match the server API doc, and second want to make it easy to verify that it matches the server API doc. When some reused type feeds into `ServerMessage` here and `ServerMessage` in `getMessages`, there's some risk to both of those goals:
- Maybe the reused type is correct for both places *now*, but the server isn't bound to keep that true forever. Also, looking closer at each of the two `ServerMessage` types against its respective API doc: even where a reused component is accurate, it's sometimes silly how imprecise it is, because we've reused it in even more places. See `ServerMessage['flags']` in either file.
- For that `ServerMessage['flags']` example, note how much effort it takes to actually decipher the intended type. In fact, it's not even explicit: we just assume that the `MESSAGE_FETCH_COMPLETE` case applies, based on our background knowledge.

On the other hand, I do know that message objects are complex to type, hmm…

Reply:
It's true there are potential trade-offs in trying to deduplicate this code.

I am interested in understanding what the differences are, though. That's partly because for each difference, there's a good chance that one or the other side is a bug. (Or potentially a problem in the API itself.) Do you know whether there are other differences?

Reply:
Makes sense; I'll list the differences below. 🙂

-----

Thinking about why I've been cautious about abstracting too far away from literal details of what we expect from the API (as in an API doc), I'm reminded of a time I *thought* I'd found a model that applied equally to two things from the server, but then we studied the API more closely and found it was wrong. In particular, a user status in the initial data, and a user status in the update-user-status event. In fact, I think I had one wrong model after another:
- First wrong model: they're both a complete user status object.
- Second wrong model: they're both an update from the then-current user status.

Rather, the _event_ holds an update from the then-current user status, and the _initial data_ holds an update from the "zero" value. See [discussion](https://github.com/zulip/zulip-mobile/pull/5224#issuecomment-1042378773) where you found a nice way to model that accurately.

And see an [instance](https://github.com/zulip/zulip-mobile/pull/5382#discussion_r940669608) where I think we might again be tempted to reach for something reusable, `UserTopic`, but it turns out to be wrong (as in my first wrong model, above).

-----

Anyway, in this case it's simpler because these are both "fetch data" endpoints that are outside the event system. I acknowledge that it really would be surprising for a server to send differently shaped messages between `GET /messages` and `GET /messages/{message_id}`, and I don't think that's likely to change from studying the API more closely.

-----

Anyway, I think these are the differences between `ServerMessage` in `getMessage` and my new `getSingleMessage` file, excluding the FL >=120 simplifications:

- The use of individual `+` vs. `$ReadOnly`, causing `typesEquivalent` to complain of one property being writable while its counterpart isn't. I think `typesEquivalent` might be wrong in these instances, but I don't know, and it's probably worth erasing these differences anyway.
- I marked `ServerReaction['user']` as write-only in `getSingleMessage`, so we don't try to read that deprecated property. This should apply to both `ServerReaction`s, I'm sure.
- I added `ServerReaction['user']['is_mirror_dummy']` in `getSingleMessage`, because I saw it in the relevant doc. Seeing now that `is_mirror_dummy` is also documented in `GET /messages`, we should probably add it in `getMessages`.
  - I made this optional in `getSingleMessage`, despite the doc saying it's required, because I observed it being absent. We could also mark it optional in the other `ServerReaction`, if we expect it to be optional there too, or if we test and find that it's sometimes absent there too, or just because optional would still be accurate, if imprecise, if it turns out to be required there.
- In both files, `ServerMessage['avatar_url']` is `string | null`.
  - In `getMessages`, IIUC, the `| null` is because we send `client_gravatar: true` to servers FL <92, and because we don't send `client_gravatar: false` for servers FL >=92.
  - In `getSingleMessage`, the server, which we know is FL >=120 since we're receiving this data, acts like it has acted for `getMessages` since FL 92—except that it doesn't actually accept a `client_gravatar` param. So the presence of `| null` there isn't conditional on anything. See [discussion](https://chat.zulip.org/#narrow/stream/378-api-design/topic/.60client_gravatar.60.20in.20.60messages.2F.7Bmessage_id.7D.60/near/1418337).

Reply:
Thanks for those details!

Yeah, it sounds like it'd be good to deduplicate these. We just discussed in the office, and a file like `src/api/rawModelTypes.js` would be a good home for `ServerMessage` and the associated types and conversions.